agent/bzero/registration/registration.go

```diff
@@ -0,0 +1,173 @@
+package registration
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/cenkalti/backoff"
+
+	logger "github.com/aws/amazon-ssm-agent/agent/log"
+)
+
+const (
+	BZeroConfigStorage    = "BZeroConfig"
+	BZeroRegErrorExitCode = 234
+
+	registrationEndpoint = "api/v2/ssm/register"
+	prodServiceUrl       = "https://cloud.bastionzero.com/" // default
+)
+
+// This is the data sent to the Reg API
+type BZeroRegRequest struct {
+	RegSecret  string `json:"registrationSecret"`
+	TargetName string `json:"instanceName"`
+	EnvId      string `json:"environmentId,omitempty"`
+	EnvName    string `json:"environmentName,omitempty"`
+}
+
+// For capturing response from registration API
+type BZeroRegResponse struct {
+	ActivationId     string `json:"activationId"`
+	ActivationCode   string `json:"activationCode"`
+	ActivationRegion string `json:"activationRegion"`
+	SSMTargetId      string `json:"ssmTargetId"`
+	OrgID            string `json:"externalOrganizationId"`
+	OrgProvider      string `json:"externalOrganizationProvider"`
+}
+
+// Attempts to register as many times as is acceptable
+func Register(log logger.T, apiKey string, envName string, envId string, targetName string, serviceUrl string) (BZeroRegResponse, error) {
+	var response BZeroRegResponse
+
+	// default target name to target hostname, if not provided
+	if targetName == "" {
+		targetName, _ = os.Hostname()
+	}
+
+	// Package data which becomes Post request body
+	var regInfo = BZeroRegRequest{
+		RegSecret:  apiKey,
+		TargetName: targetName,
+		EnvId:      envId,
+		EnvName:    envName,
+	}
+
+	// Build Registration Endpoint
+	regEndpoint := prodServiceUrl + registrationEndpoint
```

Comment:
Probably wanna use a `path.join` here instead of building it ourself, incase we change anything: 
```
u, err := url.Parse("http://foo")
u.Path = path.Join(u.Path, "bar.html")
s := u.String() // prints http://foo/bar.html
```
ref: https://stackoverflow.com/questions/34668012/combine-url-paths-with-path-join/34668130

Reply:
7193eee